File: src/lib/db/download.ts
// ************************************************
// ************** Download log stuff **************
// ************************************************

export interface RawDownloadDB {
    id: string;
    userId: string;
    dateStarted: string;
    dateEnded?: string;
    downloadURL: string;
    status: string;
    seen: string;
    success?: number;
    fail?: number;
}

export type DownloadDB<
    Keys extends keyof DownloadDBFull = keyof DownloadDBFull,
> = Pick<DownloadDBFull, Keys>;

export interface DownloadDBFull {
    id: string;
    userId: string;
    dateStarted: string;
    dateEnded?: string;
    downloadURL: string;
    status: string;
    seen: boolean;
    success?: number;
    fail?: number;
}

export const downloadQuery = `CREATE TABLE IF NOT EXISTS download (
    id TEXT NOT NULL PRIMARY KEY UNIQUE,
    userId TEXT NOT NULL,
    dateStarted TEXT NOT NULL,
    dateEnded TEXT,
    downloadURL TEXT NOT NULL,
    status TEXT NOT NULL,
    seen BOOLEAN DEFAULT "1" NOT NULL,
    success INTEGER,
    fail INTEGER
)`;

export function parseDownload(
    download: RawDownloadDB | undefined
): DownloadDB | undefined {
    if (!download) {
        return undefined;
    }

    return {
        id: download.id,
        userId: download.userId,
        dateStarted: download.dateStarted,
        dateEnded: download.dateEnded,
        downloadURL: download.downloadURL,
        status: download.status,
        seen: download.seen == "1" ? true : false,
        success: download.success,
        fail: download.fail,
    };
}

File: backend/db/download.py
from typing import TypedDict, List, Optional
import json
from dataclasses import dataclass


@dataclass
class RawDownloadDB:
    id: str
    downloadId: str
    dateStarted: str
    dateEnded: str
    downloadURL: str
    status: str
    seen: str
    success: int
    fail: int


@dataclass
class DownloadDBFull:
    id: str
    downloadId: str
    dateStarted: str
    dateEnded: str
    downloadURL: str
    status: str
    seen: bool
    success: int
    fail: int


def parse_download(raw_download: Optional[RawDownloadDB]) -> Optional[DownloadDBFull]:
    if not raw_download:
        return None

    return DownloadDBFull(
        id=raw_download.id,
        downloadId=raw_download.downloadId,
        dateStarted=raw_download.dateStarted,
        dateEnded=raw_download.dateEnded,
        downloadURL=raw_download.downloadURL,
        status=raw_download.status,
        seen=True if raw_download.seen == "1" else False,
        success=raw_download.success,
        fail=raw_download.fail
    )


download_query = """
CREATE TABLE IF NOT EXISTS download (
    id TEXT NOT NULL PRIMARY KEY UNIQUE,
    userId TEXT NOT NULL,
    dateStarted TEXT NOT NULL,
    dateEnded TEXT,
    downloadURL TEXT NOT NULL,
    status TEXT NOT NULL,
    seen BOOLEAN DEFAULT "1" NOT NULL,
    success INTEGER,
    fail INTEGER
)"""
